File: python/x402/mechanisms/tvm/exact/settlement_batcher.py
"""Internal batching helpers for TVM exact settlement relay."""

from __future__ import annotations

import queue
import threading
import time
from collections.abc import Callable
from dataclasses import dataclass, field

from ..constants import (
    DEFAULT_SETTLEMENT_BATCH_MAX_SIZE,
    DEFAULT_SETTLEMENT_CONFIRMATION_WORKERS,
    ERR_EXACT_TVM_SIMULATION_FAILED,
    ERR_EXACT_TVM_TRANSACTION_FAILED,
)
from ..settlement_cache import SettlementCache
from ..signer import FacilitatorTvmSigner
from ..types import ParsedTvmSettlement, TvmRelayRequest


@dataclass
class _BatchResult:
    success: bool
    transaction: str = ""
    error_reason: str | None = None
    error_message: str | None = None


@dataclass
class _QueuedSettlement:
    network: str
    settlement_hash: str
    settlement: ParsedTvmSettlement
    relay_request: TvmRelayRequest
    completed: threading.Event = field(default_factory=threading.Event)
    result: _BatchResult | None = None


@dataclass
class _PendingConfirmation:
    network: str
    batch: list[_QueuedSettlement]
    trace_external_hash_norm: str


class _SettlementBatcher:
    def __init__(
        self,
        signer: FacilitatorTvmSigner,
        settlement_cache: SettlementCache,
        *,
        flush_interval_seconds: float,
        batch_flush_size: int,
        confirmation_workers: int = DEFAULT_SETTLEMENT_CONFIRMATION_WORKERS,
        confirmation_timeout_seconds: float,
        settlement_verifier: Callable[[dict[str, object], ParsedTvmSettlement], str],
    ) -> None:
        self._signer = signer
        self._settlement_cache = settlement_cache
        self._flush_interval_seconds = flush_interval_seconds
        self._batch_flush_size = batch_flush_size
        self._max_batch_size = DEFAULT_SETTLEMENT_BATCH_MAX_SIZE
        self._confirmation_timeout_seconds = confirmation_timeout_seconds
        self._settlement_verifier = settlement_verifier
        if confirmation_workers < 1:
            raise ValueError("confirmation_workers must be at least 1")
        self._lock = threading.Lock()
        self._condition = threading.Condition(self._lock)
        self._confirmation_queue: queue.SimpleQueue[_PendingConfirmation] = queue.SimpleQueue()
        self._queues: dict[str, list[_QueuedSettlement]] = {}
        self._deadlines: dict[str, float] = {}
        self._worker = threading.Thread(
            target=self._run, name="tvm-settlement-batcher", daemon=True
        )
        self._worker.start()
        self._confirmation_workers = [
            threading.Thread(
                target=self._run_confirmation_worker,
                name=f"tvm-settlement-confirmation-{idx}",
                daemon=True,
            )
            for idx in range(confirmation_workers)
        ]
        for worker in self._confirmation_workers:
            worker.start()

    def enqueue(self, queued_settlement: _QueuedSettlement) -> _BatchResult:
        with self._condition:
            queue = self._queues.setdefault(queued_settlement.network, [])
            queue.append(queued_settlement)
            if len(queue) == 1:
                self._deadlines[queued_settlement.network] = (
                    time.monotonic() + self._flush_interval_seconds
                )
            elif len(queue) >= self._batch_flush_size:
                self._deadlines[queued_settlement.network] = time.monotonic()
            self._condition.notify_all()

        queued_settlement.completed.wait()
        assert queued_settlement.result is not None
        return queued_settlement.result

    def _run(self) -> None:
        while True:
            with self._condition:
                network, batch = self._wait_for_ready_batch_locked()
            self._flush_batch(network, batch)

    def _wait_for_ready_batch_locked(self) -> tuple[str, list[_QueuedSettlement]]:
        while True:
            now = time.monotonic()
            for network, deadline in list(self._deadlines.items()):
                queue = self._queues.get(network)
                if queue and deadline <= now:
                    batch_size = min(len(queue), self._max_batch_size)
                    batch = queue[:batch_size]
                    del queue[:batch_size]
                    if queue:
                        self._deadlines[network] = (
                            now
                            if len(queue) >= self._batch_flush_size
                            else now + self._flush_interval_seconds
                        )
                        self._condition.notify_all()
                    else:
                        self._queues.pop(network, None)
                        self._deadlines.pop(network, None)
                    return network, batch
            self._condition.wait(timeout=self._next_wait_timeout_locked())

    def _next_wait_timeout_locked(self) -> float | None:
        if not self._deadlines:
            return None
        return max(0.0, min(self._deadlines.values()) - time.monotonic())

    def _flush_batch(self, network: str, batch: list[_QueuedSettlement]) -> None:
        try:
            external_boc = self._signer.build_relay_external_boc_batch(
                network,
                [queued.relay_request for queued in batch],
            )
            trace_external_hash_norm = self._signer.send_external_message(network, external_boc)
        except Exception as exc:
            for queued in batch:
                self._fail_queued_settlement(
                    queued,
                    error_reason=(
                        ERR_EXACT_TVM_SIMULATION_FAILED
                        if isinstance(exc, ValueError)
                        else ERR_EXACT_TVM_TRANSACTION_FAILED
                    ),
                    error_message=str(exc),
                )
            return

        self._confirmation_queue.put(
            _PendingConfirmation(
                network=network,
                batch=batch,
                trace_external_hash_norm=trace_external_hash_norm,
            )
        )

    def _run_confirmation_worker(self) -> None:
        while True:
            pending = self._confirmation_queue.get()
            try:
                finalized_trace = self._signer.wait_for_trace_confirmation(
                    pending.network,
                    pending.trace_external_hash_norm,
                    timeout_seconds=self._confirmation_timeout_seconds,
                )
            except Exception as exc:
                for queued in pending.batch:
                    self._fail_queued_settlement(
                        queued,
                        error_reason=(
                            ERR_EXACT_TVM_SIMULATION_FAILED
                            if isinstance(exc, ValueError)
                            else ERR_EXACT_TVM_TRANSACTION_FAILED
                        ),
                        error_message=str(exc),
                    )
                continue

            for queued in pending.batch:
                try:
                    transaction_hash = self._settlement_verifier(
                        finalized_trace,
                        queued.settlement,
                    )
                    queued.result = _BatchResult(
                        success=True,
                        transaction=transaction_hash,
                    )
                except Exception as exc:
                    self._fail_queued_settlement(
                        queued,
                        error_reason=ERR_EXACT_TVM_TRANSACTION_FAILED,
                        error_message=str(exc),
                    )
                    continue
                queued.completed.set()
                # On-chain seqno already advanced; further retries fail at verify
                self._settlement_cache.release(queued.settlement_hash)

    def _fail_queued_settlement(
        self,
        queued: _QueuedSettlement,
        *,
        error_reason: str,
        error_message: str,
    ) -> None:
        queued.result = _BatchResult(
            success=False,
            transaction="",
            error_reason=error_reason,
            error_message=error_message,
        )
        queued.completed.set()
        self._settlement_cache.release(queued.settlement_hash)

File: typescript/packages/mechanisms/tvm/src/exact/settlement-batcher.ts
import {
  DEFAULT_SETTLEMENT_BATCH_FLUSH_INTERVAL_SECONDS,
  DEFAULT_SETTLEMENT_BATCH_FLUSH_SIZE,
  DEFAULT_SETTLEMENT_BATCH_MAX_SIZE,
  DEFAULT_TRACE_CONFIRMATION_TIMEOUT_SECONDS,
  ERR_EXACT_TVM_TRANSACTION_FAILED,
} from "../constants";
import type { SettlementCache } from "../settlement-cache";
import type { FacilitatorTvmSigner } from "../signer";
import type { ParsedTvmSettlement, TvmRelayRequest } from "../types";

export interface BatchResult {
  success: boolean;
  transaction?: string;
  errorReason?: string;
  errorMessage?: string;
}

export interface QueuedSettlement {
  network: string;
  settlementHash: string;
  settlement: ParsedTvmSettlement;
  relayRequest: TvmRelayRequest;
}

type PendingSettlement = QueuedSettlement & {
  resolve: (result: BatchResult) => void;
};

export class SettlementBatcher {
  private readonly queues = new Map<string, PendingSettlement[]>();
  private readonly timers = new Map<string, ReturnType<typeof setTimeout>>();

  constructor(
    private readonly signer: FacilitatorTvmSigner,
    private readonly settlementCache: SettlementCache,
    private readonly options: {
      flushIntervalSeconds?: number;
      batchFlushSize?: number;
      confirmationTimeoutSeconds?: number;
      settlementVerifier: (
        traceData: Record<string, unknown>,
        settlement: ParsedTvmSettlement,
        relayRequest: TvmRelayRequest,
      ) => string;
    },
  ) {}

  enqueue(queuedSettlement: QueuedSettlement): Promise<BatchResult> {
    return new Promise(resolve => {
      const queue = this.queues.get(queuedSettlement.network) ?? [];
      queue.push({ ...queuedSettlement, resolve });
      this.queues.set(queuedSettlement.network, queue);

      if (queue.length >= (this.options.batchFlushSize ?? DEFAULT_SETTLEMENT_BATCH_FLUSH_SIZE)) {
        this.clearTimer(queuedSettlement.network);
        void this.flushNetwork(queuedSettlement.network);
        return;
      }

      if (!this.timers.has(queuedSettlement.network)) {
        const timer = setTimeout(
          () => void this.flushNetwork(queuedSettlement.network),
          (this.options.flushIntervalSeconds ?? DEFAULT_SETTLEMENT_BATCH_FLUSH_INTERVAL_SECONDS) *
            1000,
        );
        this.timers.set(queuedSettlement.network, timer);
      }
    });
  }

  private async flushNetwork(network: string): Promise<void> {
    this.clearTimer(network);
    const queue = this.queues.get(network) ?? [];
    if (!queue.length) return;
    const batch = queue.splice(0, DEFAULT_SETTLEMENT_BATCH_MAX_SIZE);
    if (queue.length) {
      this.queues.set(network, queue);
      const timer = setTimeout(() => void this.flushNetwork(network), 0);
      this.timers.set(network, timer);
    } else {
      this.queues.delete(network);
    }

    let traceExternalHashNorm = "";
    try {
      const externalBoc = await this.signer.buildRelayExternalBocBatch(
        network,
        batch.map(item => item.relayRequest),
      );
      traceExternalHashNorm = await this.signer.sendExternalMessage(network, externalBoc);
    } catch (error) {
      this.failBatch(batch, {
        errorReason: ERR_EXACT_TVM_TRANSACTION_FAILED,
        errorMessage: error instanceof Error ? error.message : String(error),
      });
      return;
    }

    let finalizedTrace: Record<string, unknown>;
    try {
      finalizedTrace = await this.signer.waitForTraceConfirmation(network, traceExternalHashNorm, {
        timeoutSeconds:
          this.options.confirmationTimeoutSeconds ?? DEFAULT_TRACE_CONFIRMATION_TIMEOUT_SECONDS,
      });
    } catch (error) {
      this.failBatch(batch, {
        errorReason: ERR_EXACT_TVM_TRANSACTION_FAILED,
        errorMessage: error instanceof Error ? error.message : String(error),
      });
      return;
    }

    for (const queued of batch) {
      try {
        const transaction = this.options.settlementVerifier(
          finalizedTrace,
          queued.settlement,
          queued.relayRequest,
        );
        queued.resolve({ success: true, transaction });
        this.settlementCache.release(queued.settlementHash);
      } catch (error) {
        queued.resolve({
          success: false,
          transaction: "",
          errorReason: ERR_EXACT_TVM_TRANSACTION_FAILED,
          errorMessage: error instanceof Error ? error.message : String(error),
        });
        this.settlementCache.release(queued.settlementHash);
      }
    }
  }

  private failBatch(batch: PendingSettlement[], failure: Omit<BatchResult, "success">): void {
    for (const queued of batch) {
      queued.resolve({
        success: false,
        transaction: "",
        ...failure,
      });
      this.settlementCache.release(queued.settlementHash);
    }
  }

  private clearTimer(network: string): void {
    const timer = this.timers.get(network);
    if (timer) {
      clearTimeout(timer);
      this.timers.delete(network);
    }
  }
}
